Quick FYI before you spin another round:

**Lint CI is red on `7c7cdda`** — the new `'should cache per-model content generators'` test calls `client.generateContent(contents, {}, undefined, 'fast-model')` with `undefined` for the third arg, which is typed as `AbortSignal`:

```
src/core/client.test.ts(3434,50): error TS2345: Argument of type 'undefined' is not assignable to parameter of type 'AbortSignal'.
src/core/client.test.ts(3438,50): error TS2345: ...
```

`tsc --build` fails → Test job gets skipped. Pass an `AbortSignal` (e.g. `new AbortController().signal`) and it should go green.

Also one substantive thing — I think round 4 #1 got crossed wires. The "cache claims invalidation it doesn't have" item wasn't about a race condition; it's about **stale cache after settings edits**: user edits `modelProviders.fast.extra_body.enable_thinking` mid-session → same model name → cache hit → returns the pre-edit `ContentGenerator` → new setting silently ignored (the original #3765 symptom).

Switching `Map<string, ContentGenerator>` to `Map<string, Promise<ContentGenerator>>` is a nice independent improvement (avoids redundant SDK instantiation under concurrent side queries, and the `delete(model)` on failure is a good touch), but it solves the *concurrent build* problem, not the *stale entry* problem. The JSDoc line "Cleared on config changes that could affect model settings" is still describing behavior the code doesn't have — please either wire actual invalidation, or shrink the cache scope (e.g. clear on `resetChat()` / config swap) and update the JSDoc to match.

Also still outstanding from round 4: a test for `resolveModelAcrossAuthTypes` (the cross-authType fallback path — main miss → secondary hit). Currently 0 coverage on that ~30 lines.

Happy to take another careful pass once lint is green and those two are addressed.

--- a/packages/core/src/core/client.test.ts
+++ b/packages/core/src/core/client.test.ts
@@ -28,12 +28,18 @@ import { type GeminiChat } from './geminiChat.js';
 import type { Config } from '../config/config.js';
 import { ApprovalMode } from '../config/config.js';
 import type { ModelsConfig } from '../models/modelsConfig.js';
+import { retryWithBackoff } from '../utils/retry.js';
 import {
   CompressionStatus,
   GeminiEventType,
   Turn,
   type ChatCompressionInfo,
 } from './turn.js';
+
+vi.mock('../utils/retry.js', () => ({
+  retryWithBackoff: vi.fn(async (fn) => await fn()),
+  isUnattendedMode: vi.fn(() => false),
+}));
 import { getCoreSystemPrompt, getCustomSystemPrompt } from './prompts.js';
 import { DEFAULT_QWEN_FLASH_MODEL } from '../config/models.js';
 import { FileDiscoveryService } from '../services/fileDiscoveryService.js';
@@ -3387,6 +3393,9 @@ Other open files:
         apiModel: 'test-model',
       } as unknown as ContentGeneratorConfig);
 
+      // Success path for createContentGenerator
+      vi.mocked(createContentGenerator).mockResolvedValue(mockContentGenerator);
+
       await client.generateContent(
         contents,
         { temperature: 0.5 },
@@ -3394,14 +3403,40 @@ Other open files:
         'fast-model',
       );
 
-      // The retry call should receive the fast model's authType, not the main model's
-      // (In test env the retry is a no-op, but the config is passed correctly)
-      // We verify by checking createContentGeneratorForModel was exercised
-      // via the getResolvedModel call with openai authType
-      expect(getResolvedModel).toHaveBeenCalledWith(
-        AuthType.QWEN_OAUTH, // initial lookup uses main authType
-        'fast-model',
+      // VERIFY: retryWithBackoff was called with the fast model's authType ('openai'),
+      // not the main model's authType ('QWEN_OAUTH').
+      expect(retryWithBackoff).toHaveBeenCalledWith(
+        expect.any(Function),
+        expect.objectContaining({
+          authType: 'openai',
+        }),
       );
+    });
+
+    it('should cache per-model content generators', async () => {
+      const contents = [{ role: 'user', parts: [{ text: 'hello' }] }];
+      const mockResolvedModel = {
+        id: 'fast-model',
+        authType: 'openai' as const,
+        name: 'Fast Model',
+        baseUrl: 'https://fast-api.example.com',
+        generationConfig: {},
+        capabilities: {},
+      };
+
+      vi.mocked(mockConfig.getModelsConfig).mockReturnValue({
+        getResolvedModel: vi.fn().mockReturnValue(mockResolvedModel),
+      } as unknown as ModelsConfig);
+
+      vi.mocked(createContentGenerator).mockResolvedValue(mockContentGenerator);
+
+      // First call
+      await client.generateContent(contents, {}, undefined, 'fast-model');
+      expect(createContentGenerator).toHaveBeenCalledTimes(1);
+
+      // Second call - should use cache
+      await client.generateContent(contents, {}, undefined, 'fast-model');
+      expect(createContentGenerator).toHaveBeenCalledTimes(1);
     });
   });
 });

--- a/packages/core/src/core/client.test.ts
+++ b/packages/core/src/core/client.test.ts
@@ -3356,6 +3356,10 @@ Other open files:
         }),
         expect.any(String),
       );
+
+      // buildAgentContentGeneratorConfig must NOT be called when the model is
+      // not in the registry — the fallback path skips config construction.
+      expect(buildAgentContentGeneratorConfig).not.toHaveBeenCalled();
     });
 
     it('should use fast model authType for retry, not main model authType', async () => {

--- a/packages/core/src/core/client.ts
+++ b/packages/core/src/core/client.ts
@@ -146,7 +146,7 @@ export class GeminiClient {
    * on every side query (recap, title, tool summary).
    * Cleared on config changes that could affect model settings.
    */
-  private perModelGeneratorCache = new Map<string, ContentGenerator>();
+  private perModelGeneratorCache = new Map<string, Promise<ContentGenerator>>();
 
   /**
    * At any point in this conversation, was compression triggered without
@@ -1275,42 +1275,47 @@ export class GeminiClient {
   private async createContentGeneratorForModel(
     model: string,
   ): Promise<ContentGenerator> {
-    // Check cache first
+    // Check cache first (Promise coalescing to prevent redundant SDK instantiations)
     const cached = this.perModelGeneratorCache.get(model);
     if (cached) return cached;
 
-    try {
-      const resolvedModel = this.resolveModelAcrossAuthTypes(model);
+    const generatorPromise = (async () => {
+      try {
+        const resolvedModel = this.resolveModelAcrossAuthTypes(model);
 
-      if (!resolvedModel) {
-        debugLogger.warn(
-          `Model "${model}" not found in registry across all authTypes, falling back to main generator.`,
+        if (!resolvedModel) {
+          debugLogger.warn(
+            `Model "${model}" not found in registry across all authTypes, falling back to main generator.`,
+          );
+          return this.getContentGeneratorOrFail();
+        }
+
+        const targetConfig = buildAgentContentGeneratorConfig(
+          this.config,
+          model,
+          {
+            authType: resolvedModel.authType,
+            apiKey: resolvedModel.envKey
+              ? (process.env[resolvedModel.envKey] ?? undefined)
+              : undefined,
+            baseUrl: resolvedModel.baseUrl,
+          },
         );
+
+        return await createContentGenerator(targetConfig, this.config);
+      } catch (err: unknown) {
+        debugLogger.warn(
+          `Failed to create content generator for model "${model}", falling back to main generator.`,
+          err instanceof Error ? err.message : String(err),
+        );
+        // On failure, delete from cache so subsequent attempts can retry.
+        this.perModelGeneratorCache.delete(model);
         return this.getContentGeneratorOrFail();
       }
+    })();
 
-      const targetConfig = buildAgentContentGeneratorConfig(
-        this.config,
-        model,
-        {
-          authType: resolvedModel.authType,
-          apiKey: resolvedModel.envKey
-            ? (process.env[resolvedModel.envKey] ?? undefined)
-            : undefined,
-          baseUrl: resolvedModel.baseUrl,
-        },
-      );
-
-      const generator = await createContentGenerator(targetConfig, this.config);
-      this.perModelGeneratorCache.set(model, generator);
-      return generator;
-    } catch (err: unknown) {
-      debugLogger.warn(
-        `Failed to create content generator for model "${model}", falling back to main generator.`,
-        err instanceof Error ? err.message : String(err),
-      );
-      return this.getContentGeneratorOrFail();
-    }
+    this.perModelGeneratorCache.set(model, generatorPromise);
+    return generatorPromise;
   }
 
   async tryCompressChat(

--- a/packages/core/src/core/client.ts
+++ b/packages/core/src/core/client.ts
@@ -22,6 +22,7 @@ const debugLogger = createDebugLogger('CLIENT');
 
 // Core modules
 import type { ContentGenerator } from './contentGenerator.js';
+import type { ResolvedModelConfig } from '../models/types.js';
 import { AuthType, createContentGenerator } from './contentGenerator.js';
 import { GeminiChat } from './geminiChat.js';
 import {
@@ -1216,33 +1217,59 @@ export class GeminiClient {
    */
 
   /**
+   * Resolve a model across all authTypes. Handles the case where the target
+   * model is registered under a different authType than the main model
+   * (e.g. main=QWEN_OAUTH, fast=USE_ANTHROPIC).
+   *
+   * TODO: Move cross-authType resolution to ModelRegistry for a cleaner
+   * data-layer solution. Follow-up PR.
+   */
+  private resolveModelAcrossAuthTypes(
+    model: string,
+  ): ResolvedModelConfig | undefined {
+    const modelsConfig = this.config.getModelsConfig();
+    const allAuthTypes: AuthType[] = [
+      AuthType.QWEN_OAUTH,
+      AuthType.USE_OPENAI,
+      AuthType.USE_VERTEX_AI,
+      AuthType.USE_ANTHROPIC,
+      AuthType.USE_GEMINI,
+    ];
+
+    // Try the main authType first for early exit
+    const mainAuthType = this.config.getContentGeneratorConfig()?.authType;
+    if (mainAuthType) {
+      const resolved = modelsConfig.getResolvedModel(mainAuthType, model);
+      if (resolved) return resolved;
+    }
+
+    for (const authType of allAuthTypes) {
+      if (authType === mainAuthType) continue;
+      const resolved = modelsConfig.getResolvedModel(authType, model);
+      if (resolved) return resolved;
+    }
+
+    return undefined;
+  }
+
+  /**
    * Resolve the authType for a given model without creating a full generator.
    * Used by retry logic to ensure provider-specific checks (e.g. QWEN_OAUTH
    * quota detection) reference the correct provider.
    */
   private createRetryAuthTypeForModel(model: string): string | undefined {
-    const authType =
-      this.config.getContentGeneratorConfig()?.authType ?? AuthType.USE_OPENAI;
-    const resolvedModel = this.config
-      .getModelsConfig()
-      .getResolvedModel(authType, model);
-    return resolvedModel?.authType;
+    return this.resolveModelAcrossAuthTypes(model)?.authType;
   }
 
   private async createContentGeneratorForModel(
     model: string,
   ): Promise<ContentGenerator> {
     try {
-      const authType =
-        this.config.getContentGeneratorConfig()?.authType ??
-        AuthType.USE_OPENAI;
-      const resolvedModel = this.config
-        .getModelsConfig()
-        .getResolvedModel(authType, model);
+      const resolvedModel = this.resolveModelAcrossAuthTypes(model);
 
       if (!resolvedModel) {
         debugLogger.debug(
-          `Model "${model}" not found in registry for authType "${authType}", falling back to main generator.`,
+          `Model "${model}" not found in registry across all authTypes, falling back to main generator.`,
         );
         return this.getContentGeneratorOrFail();
       }

